hazelcast/src/main/java/com/hazelcast/internal/util/XmlUtil.java

```diff
@@ -111,7 +111,19 @@ public static SAXParserFactory getSAXParserFactory() throws ParserConfigurationE
      * Returns {@link XMLInputFactory} with XXE protection enabled.
      */
     public static XMLInputFactory getXMLInputFactory() {
-        XMLInputFactory xmlInputFactory = XMLInputFactory.newInstance();
+        return getProtectedFactory(XMLInputFactory.newInstance());
+    }
+
+    /**
+     * Returns {@link XMLInputFactory} with XXE protection enabled.
+     *
+     * @param factoryId name of the system property to look for factory implementation
+     */
+    public static XMLInputFactory getXMLInputFactory(String factoryId) {
+        return getProtectedFactory(XMLInputFactory.newFactory(factoryId, null));
```

Comment:
Since we want to make sure our own shaded parser is loaded, I think we have better chances by requesting load explicitly through the classloader that loaded hazelcast classes (eg `XmlUtil.class.getClassLoader()`) rather than relying on the context classloader (`null`). wdyt?

Reply:
But this goes against what you wrote here https://github.com/hazelcast/hazelcast-enterprise/pull/4607#discussion_r794258654 where we say we want to give the user freedom - I think using context classloader (with a fallback to system classloader) is most versatile.

Reply:
https://github.com/hazelcast/hazelcast/pull/20509/commits/4a6242ad5b8a5ca69fc7ebc497ef1a8266cfaadf I don't see how it whould make a difference. Though I can revert this commit if necessary.

Reply:
@frant-hartm well spotted, let's revert to context classloader. Sorry for the back-and-forth @ramizdundar .

Reply:
https://github.com/hazelcast/hazelcast/pull/20509/commits/6cfc048a7c3ac69c044b64996290ea2de057abd3